Add OP_TEMPLATEHASH and TH+CSFS+IK bundle BIP

instagibbs opened the pull request.

shesek committed:
BIP 345: Fix OP_VAULT_RECOVER specification for the recovery-sPK-hash

The recovery scriptPubKey needs to be prefixed with its CompactSize-encoded length.

darosior committed:
Add OP_TEMPLATEHASH and TH+CSFS+IK bundle BIP

nothingmuch committed:
Separate fragment params with `-` instead of `+`

Since only Bull Bitcoin Mobile and Cake wallet are currently deployed in
production, both using PDK, and the `+` character is causing some
friction, this change seems justified to avoid similar issues with
future implementations.

nothingmuch committed:
Specify lexicographical order for fragment params

The rationale for this change is that since `-` instead of `+` breaks
compatibility anyway, the marginal cost of removing this
unusual/surprising requirement for reverse lexicographical ordering is
zero.

real-or-random committed:
bip3: Switch to SPDX License Expressions

real-or-random committed:
bip3: Don't call CC0 a license

That's a bit of legal nitpicking, sorry. CC0 contains something like a
public domain dedication along with a fallback license, so it's neither
entirely. Some call it a "legal instrument". I prefer not calling it
anything.

real-or-random committed:
bip3: Don't require omitting unacceptable licenses

I think that requirement is not helpful. I don't think hat including
additional licenses will be overwhelming to the reader. If anything, it
will obfuscates the actual licensing conditions. (Anyway, this should be
super rare.)

real-or-random committed:
bip3: Recommend SPDX-License-Identifier comments

real-or-random committed:
bip3: Editorial changes in "BIP Licensing"

real-or-random committed:
bip3: Change License example to CC0-1.0 OR MIT

The actual reason why I suggest this is that I think that's a great
default choice for a new BIP, so it's a perfect example. CC0-1.0 is a
great liberal choice for the BIP document (and test vectors etc.), and
MIT is the common choice for code in our ecosystem. Putting both BIP and
code under the "OR" avoids any confusion about which part is licensed
under which terms and also avoids any hassle when reorganizing, e.g.,
when moving code out of the BIP Markdown file to a separate file etc.

But I don't want this PR to recommend a license, so let me sell this
change as an editorial change to an example, which is warranted because
the MIT is much more known than FSFAP, in particular in this ecosystem.

real-or-random committed:
bip3: Fix SPDX id of FSF/GNU All Permissive

real-or-random committed:
bip3: Fix SPDX id of Open Publication License

real-or-random committed:
bip3: Use user-defined LicenseRef-PD instead of PD

SPDX doesn't have an official identifier for "public domain", at least
not for the simple "This document is placed into the public domain"
declarations used in some BIPs, see
https://wiki.spdx.org/view/Legal_Team/Decisions/Dealing_with_Public_Domain_within_SPDX_Files
for the rationale provided by their legal team. The rationale is sound,
but It's possible to create "user-defined" identifiers of the form
LicenseRef-X. This is a good idea here to make sure that all SPDX
expression will be formally valid.

And in our case, all "PD" BIPs match the following pseudo regex, so
there's not much potential for confusion:

    "This (document|BIP|work|proposal) is (hereby)? (placed)? in the
    public domain."

So it makes sense to keep using a single identifier for all of these.

real-or-random committed:
bip3: Editorial cleanup of the license lists

Ademan commented on bip-templatehash.md:
```diff
@@ -0,0 +1,171 @@
+several important ways.
+
+First of all, `OP_TEMPLATEHASH` is only defined for Tapscript, as modifying legacy Script comes with an unnecessarily
+increased risk surface. In addition, sticking to Tapscript allows to leverage more powerful upgrade hooks (`OP_SUCCESS`s
```
nit:

s/allows to leverage/allows leveraging/

Not sure if that phrasing is intentional but it's awkward to my eyes.

instagibbs replied:
taken

Ademan commented on bip-templatehash.md:
```diff
@@ -0,0 +1,171 @@
+techniques unless additional signatures are included, as used for instance [in the LN-Symmetry demo][symmetry-annex-publication].
+
+Programmable transaction introspection capabilities have been proposed as an alternative to a primitive which only
+allows to commit to the exact next transaction. It remains to be shown that these more flexible capabilities do
```
nit: s/allow to commit/allows committing/

same as previous

instagibbs replied:
taken

Ademan commented on bip-templatehash.md:
```diff
@@ -0,0 +1,171 @@
+## Acknowledgements
+
+Credit to Jeremy Rubin for his leadership and perseverance in defending how a simple primitive
+allowing to commit to the entire spending transaction is useful for reducing
```
same as above except I'm not sure how to reword it *well*

s/simple primitive allowing to commit/simple primitive which allows committing/

maybe?

instagibbs replied:
modified

nothingmuch committed:
CI: Enforce BIP 2 & 3 field ordering requirements

The specified field order is consistent with both BIPs 2 and 3. The
ordering of fields which are only present in one or the other is
ambiguous, e.g. as in `Proposed-Replacement` and `Superseded-By` but
only one of these applies to a given BIP.

The `Editor` field is spurious, only being used in BIP 69, and appears
after Author.

nothingmuch committed:
scripted-diff: fix BIP 2 field order violations

-BEGIN VERIFY SCRIPT-
set -e
perl <<'-END PERL-'
use strict;
use warnings;

my $topbip = 9999;

my @FieldOrder = qw(
	BIP
	Layer
	Title
	Author
	Authors
	Editor
	Deputies
	Discussions-To
	Comments-Summary
	Comments-URI
	Status
	Type
	Created
	License
	License-Code
	Discussion
	Post-History
	Version
	Requires
	Replaces
	Proposed-Replacement
	Superseded-By
);

my $bipnum = 0;
while (++$bipnum <= $topbip) {
	my $fn = sprintf "bip-%04d.mediawiki", $bipnum;
	my $is_markdown = 0;
	if (!-e $fn) {
		$fn = sprintf "bip-%04d.md", $bipnum;
		$is_markdown = 1;
	}
	-e $fn || next;
	open my $F, "<", $fn or die "$!";

	my (@before, %preamble, @after);

	if ($is_markdown) {
		while (<$F>) {
			push @before, $_;
			last if m[^(?:\xef\xbb\xbf)?```$]
		}
		die "No ``` in $fn" if eof $F;
	} else {
		while (<$F>) {
			push @before, $_;
			last if m[^(?:\xef\xbb\xbf)?<pre>$];
		}
		die "No <pre> in $fn" if eof $F;
	}
	my %found;
	my ($title, $author, $status, $type, $layer);
	my ($field, $val, @field_order);
	while (<$F>) {
		push @after, $_ and last if ($is_markdown && m[^```$]);
		push @after, $_ and last if (!$is_markdown && m[^</pre>$]);

		if (m[^  ([\w-]+)\: (.*\S)$]) {
			$field = $1;
			$val = $2;
		} elsif (m[^  ( +)(.*\S)$]) {
			$val = $2;
		} else {
			die "Bad line in $fn preamble";
		}

		push @{$preamble{$field} ||= []}, $_;
	}
	push @after, <$F>;
	close $F or die $!;

	open my $W, ">", "$fn" or die "$!";

	print $W @before;
	print $W map { @$_ } grep { defined } delete @preamble{@FieldOrder};
	die "Unknown fields: @{[ keys %preamble ]}" if %preamble;
	print $W @after;

	close $W or die $!;
}
-END PERL-
-END VERIFY SCRIPT-

murchandamus committed:
Merge pull request #1890 from nothingmuch/fragment-fixes

BIP 77: Delimit fragment params with  `-` instead of `+`

Galoretka committed:
Fix function name typo: use bytearr_cmp instead of bytearray_cmp

murchandamus committed:
Merge pull request #1897 from Galoretka/rawe

BIP69: fix function name typo in example code

jonatack committed:
Merge pull request #1892 from real-or-random/202507-spdx

bip3: Switch to SPDX identifiers

dplusplus1024 committed:
Fix: Use plural “bytes” in serialization table to match field size convention

Aligns the first field in the serialization table with the rest, which use plural “bytes” (e.g., “4 bytes: child number”) for consistency.

jonatack committed:
Merge branch 'nothingmuch-field-ordering-pr'

* nothingmuch-field-ordering-pr:
  CI: Enforce BIP 2 & 3 field ordering requirements
  scripted-diff: fix BIP 2 field order violations

crStiv committed:
Update bip-0077.md

jonatack committed:
Merge pull request #1901 from crStiv/typo

BIP77: fix links to Client/Directory interactions

josibake committed:
doc: be explicit for the input_hash corner case

jonatack committed:
Merge pull request #1902 from josibake/bip352-updates

BIP352: be explicit for the input_hash corner case

jonatack committed:
BIP3: drop optional License Code header

Co-authored-by: Tim Ruffing <crypto@timruffing.de>
Co-authored-by: Murch <murch@murch.one>

jonatack committed:
BIP3: refer to CC0 1.0 Universal

per https://creativecommons.org/publicdomain/zero/1.0/legalcode.en

and add a missing word

Co-authored-by: Tim Ruffing <crypto@timruffing.de>

murchandamus committed:
Merge pull request #1899 from dplusplus1024/patch-7

BIP32: Use plural “bytes”

otc-png committed:
remove repeated "the"

GarmashAlex committed:
fix broken link

murchandamus committed:
Merge pull request #1904 from jonatack/2025-07-remove-License-Code-from-BIP3

BIP3: drop optional License Code header

anim001k committed:
Update bip-0070.mediawiki

jonatack committed:
Merge pull request #1906 from GarmashAlex/links

Fix broken MES16.pdf link in BIP-347

jonatack committed:
Merge pull request #1907 from anim001k/master

Fix dead link in BIP-70: Update Mozilla root store URL

jonatack committed:
Merge pull request #1905 from otc-png/patch-1

[bip-0010]: remove repeated "the"

bethoffman committed:
Update bip-0443.mediawiki

josibake committed:
BIP352: add warning against omitting outputs (#1908)

* doc: add warning against omitting outputs

While implied by the specification, add an explicit warning that
generated outputs MUST not be omitted from the final transaction.

Co-authored-by: Mark "Murch" Erhardt <murch@murch.one>

murchandamus committed:
Merge pull request #1910 from bethoffman/fix-typo

BIP443: fix typo in the section "Transaction-wide initialization"

Forostovec committed:
fix: broken link to payment request generator

jonatack committed:
Merge pull request #1797 from shesek/patch-3

murchandamus committed:
BIP3: Update License header format on activation

Addresses follow-up from #1892

jonatack committed:
Merge pull request #1918 from murchandamus/bip3-reformat-existing-licenses

BIP3: Update License header format on activation

strmfos committed:
BIP374: remove debug print statement from BIP-374 test runner

jonatack committed:
Merge pull request #1920 from strmfos/master

BIP374: remove debug print from test runner

radik878 committed:
test(reference): pass message when verifying tampered proof

jonatack committed:
ci: update github actions checkout version

see https://github.com/actions/checkout/releases/tag/v5.0.0

jonatack committed:
Merge pull request #1914 from Forostovec/fix/broken-link

BIP 70: fix link to payment request generator

Commit:
BIP3: fix license URLs with 301 redirects

ANtutov committed:
bip-0345: fix broken link

jonatack committed:
Merge pull request #1929 from jonatack/2025-08-ci-update

ci: update github actions checkout version

jonatack committed:
BIP155: update Draft status to Final

BIP155 was deployed in Bitcoin Core version v0.21.0, and has been in use for
almost 5 years.

New networks may be added to the reserved network IDs table when they're
needed, either in this BIP or in a new one.

If BIP3 is activated, I think BIP155 would become Deployed.

Co-authored-by: Murch <murch@murch.one>
Co-authored-by: laanwj <126646+laanwj@users.noreply.github.com>

strmfos committed:
scripts: remove unused FILES variable in link-format-chk.sh

jonatack committed:
Merge pull request #1932 from ANtutov/broken-link

BIP345: fix broken link

jonatack committed:
Merge pull request #1934 from strmfos/master

scripts: remove unused FILES variable in link-format-chk.sh

Fibonacci747 committed:
feat: remove unused permutations import from bip-0352/reference.py

jonatack committed:
Merge pull request #1935 from Fibonacci747/unused-import

BIP352: remove unused import from reference.py

sashass1315 committed:
BIP345: fix broken links (#1938)

* BIP345: fix broken links

Commit:
BIP3: add MIT-0 to acceptable licenses

jonatack committed:
BIP3: update MIT license name

per both of:

- https://opensource.org/license/MIT
- https://spdx.org/licenses/MIT

jonatack committed:
BIP2: update MIT license name

per both of:

- https://opensource.org/license/MIT
- https://spdx.org/licenses/MIT

Co-authored-by: Murch <murch@murch.one>

jonatack committed:
BIP2: fix 301 redirects in license URLs

Co-authored-by: 1BitcoinBoWP1FZ4xwTNkq6XksKidmgYYw <contact@bitcoin.foundation>

ethicnology committed:
fix: replace Base64 by Base85 encoding in PWD BASE85 section

ethicnology committed:
refactor: unify capitalization

jonatack committed:
Merge pull request #1939 from ethicnology/master

BIP85: replace Base64 by Base85 in PWD BASE85 section

murchandamus committed:
Merge pull request #1940 from jonatack/2025-08-bip3-licenses

BIP3: license section updates

murchandamus committed:
Merge pull request #1941 from jonatack/2025-08-bip2-licenses

BIP2: license section fixups

aso20455 committed:
fix: correct variable assignment in bytes_to_point function

murchandamus committed:
Merge pull request #1933 from jonatack/2025-08-update-BIP155-status

BIP155: update status from Draft to Final

jonatack committed:
Merge pull request #1947 from aso20455/master

BIP328: fix assignment in bytes_to_point function

darosior committed:
bip54: fix off-by-one in creation date

jonatack committed:
Merge pull request #1950 from darosior/2509_bip54_fix_date

bip54: fix off-by-one in creation date

MozirDmitriy committed:
fix(wallet_policies): use descriptor instead of desc to prevent NameError

macgyver13 committed:
add intermediate vector material, validate added material in reference tests

jonatack committed:
Merge pull request #1952 from MozirDmitriy/mzd

BIP388: fix variable name in from_descriptor() to prevent NameError

macgyver13 committed:
address feedback (newline + extra whitespace)

jonatack committed:
Merge pull request #1953 from macgyver13/bip352-generate-intermediate-comp

BIP352: Add intermediate vector material for silent payments

jonatack committed:
BIP111: update status from Proposed to Final

murchandamus committed:
Merge pull request #1956 from jonatack/2025-09-update-BIP111-status-from-Proposed-to-Final

BIP111: update status from Proposed to Final

jonatack committed:
BIP157: update status from Draft to Final

jonatack committed:
BIP158: update status from Draft to Final

jonatack committed:
BIPs 157, 158: add "Requires" headers to each other

murchandamus committed:
Merge pull request #1963 from jonatack/2025-09_update_BIPs_157-158_to_final

BIPs 157, 158: update status to Final, add Requires header

jonatack committed:
Merge pull request #1926 from radik878/test/dleq-pass-message-in-tampered-proof

BIP374: in tests, pass message when verifying proof with message

instagibbs commented:
opened https://github.com/bitcoin/bips/pull/1974
